java-checks/src/main/java/org/sonar/java/checks/security/DataHashingCheck.java

```diff
@@ -19,9 +19,33 @@
  */
 package org.sonar.java.checks.security;
 
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
 import org.sonar.check.Rule;
 import org.sonar.java.checks.AbstractHashAlgorithmChecker;
 
 @Rule(key = "S4790")
 public class DataHashingCheck extends AbstractHashAlgorithmChecker {
+
+  public static final Set<String> DEPRECATED_HASH_CLASSES = ImmutableSet.of(
+    DeprecatedSpringPasswordEncoder.MD5.classFqn,
+    DeprecatedSpringPasswordEncoder.SHA.classFqn,
+    DeprecatedSpringPasswordEncoder.LDAP.classFqn,
+    DeprecatedSpringPasswordEncoder.MD4.classFqn,
+    DeprecatedSpringPasswordEncoder.MESSAGE_DIGEST.classFqn,
+    DeprecatedSpringPasswordEncoder.NO_OP.classFqn,
+    DeprecatedSpringPasswordEncoder.STANDARD.classFqn
+  );
+
+  private static final String MESSAGE = "Make sure this weak hash algorithm is not used in a sensitive context here.";
+
+  @Override
+  protected String getMessageForClass(String className) {
+    return DEPRECATED_HASH_CLASSES.contains(className) ? MESSAGE : null;
+  }
```

Comment:
The result of this method is used without null check, relying on the fact that `DEPRECATED_HASH_CLASSES` set contains exactly all the value of `DeprecatedSpringPasswordEncoder`, synchronization that has to be done "by hand". I'm not sure it's a good pattern, I would rather return an Optional or consistently check it for null here.

Reply:
I refactored it, but to my mind, as far as the implementation of AbstractMethodDetection and AbstractHashAlgorithmChecker guarantees there will never happen null. I someone will need to add a new class and forget to update it, there will be problem during implementation and testing. So will never go to production.

On the other hand, if we just hide someone's mistake by adding optional, that might not be a good experience to fix it.